src/Microsoft.ML.Data/Prediction/Calibrator.cs

```diff
@@ -116,21 +116,44 @@ internal interface ISelfCalibratingPredictor
         IPredictor Calibrate(IChannel ch, IDataView data, ICalibratorTrainer caliTrainer, int maxRows);
     }
 
-    public abstract class CalibratedPredictorBase :
+    /// <summary>
+    /// <see cref="IHasWeeklyTypedCalibratedPredictor"/> provides a weekly-typed way to access strongly-typed
+    /// <see cref="CalibratedPredictorBase{TSubPredictor, TCalibrator}.SubPredictor"/> and
+    /// <see cref="CalibratedPredictorBase{TSubPredictor, TCalibrator}.Calibrator"/>.
+    /// <see cref="IHasWeeklyTypedCalibratedPredictor"/> is commonly used in weekly-typed expressions. The
+    /// existence of this interface is just for supporting existing codebase, so we discourage its uses.
+    /// </summary>
+    [BestFriend]
+    internal interface IHasWeeklyTypedCalibratedPredictor
+    {
+        IPredictorProducing<float> WeeklyTypedSubPredictor { get; }
+        ICalibrator WeeklyTypedCalibrator { get; }
+    }
+
+    public abstract class CalibratedPredictorBase<TSubPredictor, TCalibrator> :
         IDistPredictorProducing<float, float>,
         ICanSaveInIniFormat,
         ICanSaveInTextFormat,
         ICanSaveInSourceCode,
         ICanSaveSummary,
-        ICanGetSummaryInKeyValuePairs
+        ICanGetSummaryInKeyValuePairs,
+        IHasWeeklyTypedCalibratedPredictor
+        where TSubPredictor : class, IPredictorProducing<float>
+        where TCalibrator : class, ICalibrator
     {
         protected readonly IHost Host;
 
-        public IPredictorProducing<float> SubPredictor { get; }
-        public ICalibrator Calibrator { get; }
+        // Strongly-typed members.
+        public TSubPredictor SubPredictor { get; }
```

Comment:
>SubPredictor [](start = 29, length = 12)

Note that we will not want this name in future, since we have deprecated the phrase "predictor" (since this style of object does not predict and has not for years). I do not suggest changing the name in this PR necessary.

Reply:
Sounds good I will do the following replacements.
`SubPredictor` ---> `SubModelParameters` 
`TSubPredictor` ---> `TSubModelParameters`

---
In reply to: [254541699](https://github.com/dotnet/machinelearning/pull/2435#discussion_r254541699) [](ancestors = 254541699)

Reply:
Maybe. However, one thing that occurred to me is that in the other prominent place where we expose the model parameters, we don't call them `ModelParameters` but just `Model`. I might elect to be consistent with that usage there. (So, `SubModel` for the property name, and `TSubModel` for the predictor.)

https://github.com/dotnet/machinelearning/blob/834e4715afa070f5e5c3628693682943b3f0bd6d/src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs#L43

Reply:
Ah. Got it. Will scan through my changes to make it.

---
In reply to: [255158805](https://github.com/dotnet/machinelearning/pull/2435#discussion_r255158805) [](ancestors = 255158805)

Reply:
I was just thinking about the same thing looking at the test code! I think it's a very good idea.

---
In reply to: [255162189](https://github.com/dotnet/machinelearning/pull/2435#discussion_r255162189) [](ancestors = 255162189,255158805)